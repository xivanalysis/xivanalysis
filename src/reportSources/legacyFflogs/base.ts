import {EventActor} from 'fflogs'
import {Actor} from 'report'

export function resolveActorId(opts: {
	id?: number | string,
	instance?: number,
	actor?: EventActor,
}): Actor['id'] {
	const idNum = (opts.id ?? opts.actor?.id ?? -1)
	const id = idNum === -1 ? 'unknown' : idNum.toString()
	const instance = opts.instance ?? 1
	return instance > 1
		? `${id}:${instance}`
		: id
}
