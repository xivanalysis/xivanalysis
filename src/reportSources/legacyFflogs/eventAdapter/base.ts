import {Event} from 'event'
import {EventActor, FflogsEvent} from 'fflogs'
import {Actor, Pull, Report} from 'report'

export function resolveActorId(opts: {
	id?: number,
	instance?: number,
	actor?: EventActor
}): Actor['id'] {
	const idNum = (opts.id ?? opts.actor?.id ?? -1)
	const id = idNum === -1 ? 'unknown' : idNum.toString()
	const instance = opts.instance ?? 1
	return instance > 1
		? `${id}:${instance}`
		: id
}

export const PREPULL_OFFSETS = {
	STATUS_ACTION: -3,
	STATUS_APPLY: -2,
	PULL_ACTION: -1,
}

// This stuff will probably be moved to a shared location for other sources to use

export interface AdapterOptions {
	report: Report
	pull: Pull
}

// TODO: Remove this once the legacy system is phased out - we only need explicit mutation handling
// while the legacy system is actually using these events
export interface MutationAdaptionResult {
	dangerouslyMutatedBaseEvent: FflogsEvent
	adaptedEvents: Event[]
}

export abstract class AdapterStep {
	protected report: Report
	protected pull: Pull

	constructor(opts: AdapterOptions) {
		this.report = opts.report
		this.pull = opts.pull
	}

	/**
	 * Perform logic to adapt a report source event into zero or more xiva events.
	 * This will be called once with each report source event, in the order provided
	 * by the source.
	 */
	adapt(baseEvent: FflogsEvent, adaptedEvents: Event[]): Event[] | MutationAdaptionResult {
		return adaptedEvents
	}

	/**
	 * Perform postprocessing on the final array of adapted events. This will be
	 * called once after all report source events have been adapted.
	 */
	postprocess(adaptedEvents: Event[]): Event[] {
		return adaptedEvents
	}
}
