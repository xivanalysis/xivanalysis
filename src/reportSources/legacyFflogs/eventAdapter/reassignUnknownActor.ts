import {Event} from 'event'
import {FflogsEvent} from 'fflogs'
import {Actor} from 'report'
import {AdapterStep, resolveActorId} from './base'

/**
 * FFLogs doesn't seem to guarantee that every actor in events is present in
 * the report the events are for. We're patching around this by checking the
 * source and target of the events, and patching them to -1 if we don't see it
 * in the report. -1 mapped in the translate step to a special sentinel actor.
 *
 * Realistically, this will primarily be caused by inlined actors. We don't
 * currently account for these - not really worth it, as they'll never be
 * primary targets or players.
 */
export class ReassignUnknownActorStep extends AdapterStep {
	// Optimisation; store results of ID lookups to save looping the actors array every time
	private knownIds = new Map<Actor['id'], boolean>()

	adapt(baseEvent: FflogsEvent, adaptedEvents: Event[]): Event[] {
		this.reassign(baseEvent, 'source')
		this.reassign(baseEvent, 'target')

		return adaptedEvents
	}

	private reassign<F extends 'source' | 'target'>(event: FflogsEvent, field: F) {
		const idField = `${field}ID` as const
		const instanceField = `${field}Instance` as const

		const actorId = resolveActorId({
			id: event[idField],
			instance: event[instanceField],
			actor: event[field],
		})

		let exists = this.knownIds.get(actorId)
		if (exists == null) {
			exists = this.pull.actors.some(actor => actor.id === actorId)
			this.knownIds.set(actorId, exists)
		}
		if (exists) { return }

		// Am I seriously mutating the fflogs events?
		// Yup. This is cursed code.
		event[idField] = -1
		delete event[instanceField]
		delete event[field]
	}
}
