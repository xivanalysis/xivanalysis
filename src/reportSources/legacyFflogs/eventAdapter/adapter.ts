import {Event} from 'event'
import {FflogsEvent} from 'fflogs'
import {sortEvents} from 'parser/core/EventSorting'
import {Pull, Report} from 'report'
import {AdapterOptions, AdapterStep} from './base'
import {DeduplicateActorUpdateStep} from './deduplicateActorUpdates'
import {DeduplicateStatusApplicationStep} from './deduplicateStatus'
import {PrepullActionAdapterStep} from './prepullAction'
import {PrepullStatusAdapterStep} from './prepullStatus'
import {ReassignUnknownActorStep} from './reassignUnknownActor'
import {TranslateAdapterStep} from './translate'

/** Adapt an array of FFLogs APIv1 events to xiva representation. */
export function adaptEvents(report: Report, pull: Pull, events: FflogsEvent[]): Event[] {
	const adapter = new EventAdapter({report, pull})

	// TODO: Move sort logic into adapter scope once legacy is removed
	sortEvents(events)

	return adapter.adaptEvents(events)
}

class EventAdapter {
	private adaptionSteps: AdapterStep[]

	constructor(opts: AdapterOptions) {
		this.adaptionSteps = [
			new ReassignUnknownActorStep(opts),
			new TranslateAdapterStep(opts),
			new DeduplicateStatusApplicationStep(opts),
			new DeduplicateActorUpdateStep(opts),
			new PrepullActionAdapterStep(opts),
			new PrepullStatusAdapterStep(opts),
		]
	}

	adaptEvents(events: FflogsEvent[]): Event[] {
		const adaptedEvents = events.flatMap(baseEvent =>
			this.adaptionSteps.reduce(
				(adaptedEvents, step) => step.adapt(baseEvent, adaptedEvents),
				[] as Event[]
			)
		)

		return this.adaptionSteps.reduce(
			(processedEvents, step) => step.postprocess(processedEvents),
			adaptedEvents,
		)
	}
}
