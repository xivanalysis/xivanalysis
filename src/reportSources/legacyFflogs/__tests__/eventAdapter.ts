import {GameEdition} from 'data/EDITIONS'
import {Events} from 'event'
import {FflogsEvent, HitType, ReportLanguage} from 'fflogs'
import {Actor, Pull, Report, Team} from 'report'
import {adaptEvents} from '../eventAdapter'
import {AdapterStep} from '../eventAdapter/base'
import {ReassignUnknownActorStep} from '../eventAdapter/reassignUnknownActor'

// "Mock" the reassign unknown actor step with its real implementation. We use this mock handling later
// to disable the step on a test-by-test basis.
// TODO: If we need to do this >1 times, work out a cleaner way of doing this.
jest.mock('../eventAdapter/reassignUnknownActor')
type ReassignUnknownActorStepParams = ConstructorParameters<typeof ReassignUnknownActorStep>
const MockReassignUnknownActorStep = ReassignUnknownActorStep as jest.Mock<ReassignUnknownActorStep>
MockReassignUnknownActorStep.mockImplementation((...args: ReassignUnknownActorStepParams) => {
	const {ReassignUnknownActorStep} = jest.requireActual('../eventAdapter/reassignUnknownActor')
	return new ReassignUnknownActorStep(...args)
})

const actor: Actor = {
	id: '1',
	kind: '1',
	name: 'Test Actor',
	team: Team.FRIEND,
	playerControlled: true,
	job: 'UNKNOWN',
}

const pull: Pull = {
	id: '1',
	timestamp: 0,
	duration: 1,
	encounter: {
		name: 'test encounter',
		duty: {
			id: 1,
			name: 'test duty',
		},
	},
	actors: [actor],
}

const report: Report = {
	timestamp: 0,
	edition: GameEdition.GLOBAL,
	name: 'Event adapter test',
	pulls: [pull],
	meta: {
		source: 'legacyFflogs',
		code: 'adapterTest',
		loading: false,
		enemies: [],
		enemyPets: [],
		fights: [],
		friendlies: [],
		friendlyPets: [],
		lang: ReportLanguage.ENGLISH,
		phases: [],
		end: Infinity,
		owner: 'test',
		start: 0,
		title: 'Event adapter test',
		zone: 0,
	},
}

// Below is a "fake" definition of every fflogs type we have typed in the codebase.
// It's long - you'll probably want to collapse it
// #region FFLogs event definitions

const fakeAbility = {
	name: 'Fake Ability',
	guid: -1,
	type: 0,
	abilityIcon: 'fakeAbilityIcon',
}

// For stuff that will never have them
const fakeBaseFields = {
	sourceIsFriendly: false,
	targetIsFriendly: false,
}

// These are injected by the hit type module, I can't be bothered typing around them
const fakeHitTypeFields = {
	criticalHit: false,
	directHit: false,
	successfulHit: true,
}

const fakeEvents: Record<FflogsEvent['type'], FflogsEvent> = {
	encounterstart: {
		...fakeBaseFields,
		timestamp: 10131563,
		type: 'encounterstart',
		encounterID: 4525,
		name: "the Heroes' Gauntlet",
		difficulty: 10,
		size: 4,
		level: 0,
		affixes: [],
	},
	encounterend: {
		...fakeBaseFields,
		timestamp: 8588978,
		type: 'encounterend',
		encounterID: 947,
		name: "Eden's Promise: Litany (Savage)",
		difficulty: 0,
		size: 8,
		kill: true,
	},
	calculateddamage: {
		timestamp: 8011183,
		type: 'calculateddamage',
		sourceID: 1,
		sourceIsFriendly: true,
		targetID: 3,
		targetIsFriendly: false,
		ability: {
			name: 'Fang and Claw',
			guid: 3554,
			type: 128,
			abilityIcon: '002000-002582.png',
		},
		hitType: 2,
		amount: 0,
		debugMultiplier: 1.1,
		packetID: 16647,
		sourceResources: {
			hitPoints: 142411,
			maxHitPoints: 142411,
			mp: 10000,
			maxMP: 10000,
			tp: 0,
			maxTP: 1000,
			x: 9260,
			y: 10463,
			facing: -245,
			absorb: 0,
		},
		targetResources: {
			hitPoints: 67180802,
			maxHitPoints: 75722720,
			mp: 10000,
			maxMP: 10000,
			tp: 0,
			maxTP: 1000,
			x: 9999,
			y: 10081,
			facing: -160,
			absorb: 0,
		},
		...fakeHitTypeFields,
	},
	damage: {
		ability: {
			abilityIcon: '012000-012507.png',
			guid: 1000725,
			name: 'Goring Blade',
			type: 1,
		},
		absorbed: 0,
		actorPotencyRatio: 35.96633490558984,
		amount: 4176,
		directHitPercentage: 0.07666098807495741,
		expectedAmount: 2980,
		expectedCritRate: 256,
		finalizedAmount: 3508.2779110051106,
		guessAmount: 3057.1384669751365,
		hitType: 1,
		multiplier: 1,
		simulated: true,
		sourceID: 2,
		sourceIsFriendly: true,
		targetID: 3,
		targetIsFriendly: false,
		targetResources: {
			absorb: 0,
			facing: -160,
			hitPoints: 66978953,
			maxHitPoints: 75722720,
			maxMP: 10000,
			maxTP: 1000,
			mp: 10000,
			tp: 0,
			x: 9999,
			y: 10081,
		},
		tick: true,
		timestamp: 8012166,
		type: 'damage',
		...fakeHitTypeFields,
	},
	calculatedheal: {
		timestamp: 7412071,
		type: 'calculatedheal',
		sourceID: 3,
		sourceIsFriendly: true,
		targetID: 3,
		targetIsFriendly: true,
		ability: {
			name: 'Assize',
			guid: 3571,
			type: 1024,
			abilityIcon: '002000-002634.png',
		},
		hitType: 2,
		amount: 36194,
		packetID: 4570,
		sourceResources: {
			hitPoints: 122214,
			maxHitPoints: 122214,
			mp: 9800,
			maxMP: 10000,
			tp: 0,
			maxTP: 1000,
			x: 9887,
			y: 10642,
			facing: -208,
			absorb: 0,
		},
		targetResources: {
			hitPoints: 122214,
			maxHitPoints: 122214,
			mp: 9800,
			maxMP: 10000,
			tp: 0,
			maxTP: 1000,
			x: 9887,
			y: 10642,
			facing: -208,
			absorb: 0,
		},
		...fakeHitTypeFields,
	},
	heal: {
		timestamp: 8586700,
		type: 'heal',
		sourceID: 11,
		sourceIsFriendly: true,
		targetID: 8,
		targetIsFriendly: true,
		ability: {
			name: 'Combined HoTs',
			guid: 500001,
			type: 1,
			abilityIcon: '000000-000405.png',
		},
		hitType: 1,
		amount: 4588,
		tick: true,
		targetResources: {
			hitPoints: 60545,
			maxHitPoints: 128026,
			mp: 9461,
			maxMP: 10000,
			tp: 0,
			maxTP: 1000,
			x: 10601,
			y: 9971,
			facing: -288,
			absorb: 0,
		},
		...fakeHitTypeFields,
	},
	begincast: {
		timestamp: 7446878,
		type: 'begincast',
		sourceID: 9,
		sourceIsFriendly: true,
		targetID: 2,
		targetIsFriendly: false,
		ability: {
			name: 'Broil III',
			guid: 16541,
			type: 1024,
			abilityIcon: '002000-002821.png',
		},
	},
	cast: {
		timestamp: 7537601,
		type: 'cast',
		sourceID: 4,
		sourceIsFriendly: true,
		targetID: 2,
		targetIsFriendly: false,
		ability: {
			name: 'Brutal Shell',
			guid: 16139,
			type: 128,
			abilityIcon: '003000-003403.png',
		},
	},
	applybuff: {
		timestamp: 7537469,
		type: 'applybuff',
		sourceID: 11,
		sourceInstance: 2,
		sourceIsFriendly: true,
		targetID: 4,
		targetIsFriendly: true,
		ability: {
			name: 'Seraphic Veil',
			guid: 1001917,
			type: 8,
			abilityIcon: '012000-012848.png',
		},
	},
	applydebuff: {
		timestamp: 190177,
		type: 'applydebuff',
		sourceID: 6,
		sourceIsFriendly: true,
		targetID: 11,
		targetIsFriendly: false,
		ability: {
			name: 'Vulnerability Up',
			guid: 1000638,
			type: 1,
			abilityIcon: '015000-015020.png',
		},
	},
	refreshbuff: {
		timestamp: 7416935,
		type: 'refreshbuff',
		sourceID: 3,
		sourceIsFriendly: true,
		targetID: 3,
		targetIsFriendly: true,
		ability: {
			name: 'Temperance',
			guid: 1001873,
			type: 1,
			abilityIcon: '012000-012633.png',
		},
	},
	refreshdebuff: {
		timestamp: 7423728,
		type: 'refreshdebuff',
		sourceID: 8,
		sourceIsFriendly: true,
		targetID: 2,
		targetIsFriendly: false,
		ability: {
			name: 'Wildfire',
			guid: 1000861,
			type: 1,
			abilityIcon: '013000-013011.png',
		},
	},
	removebuff: {
		timestamp: 7735462,
		type: 'removebuff',
		sourceID: 1,
		sourceIsFriendly: true,
		targetID: 1,
		targetIsFriendly: true,
		ability: {
			name: 'Sword Oath',
			guid: 1001902,
			type: 1,
			abilityIcon: '019000-019461.png',
		},
	},
	removedebuff: {
		timestamp: 7427662,
		type: 'removedebuff',
		sourceID: 9,
		sourceIsFriendly: true,
		targetID: 2,
		targetIsFriendly: false,
		ability: {
			name: 'Chain Stratagem',
			guid: 1001221,
			type: 1,
			abilityIcon: '012000-012809.png',
		},
	},
	applybuffstack: {
		timestamp: 7410286,
		type: 'applybuffstack',
		sourceID: 7,
		sourceIsFriendly: true,
		targetID: 7,
		targetIsFriendly: true,
		ability: {
			name: 'Acceleration',
			guid: 1001238,
			type: 1,
			abilityIcon: '019000-019661.png',
		},
		stack: 2,
	},
	applydebuffstack: {
		timestamp: 7425207,
		type: 'applydebuffstack',
		sourceID: 8,
		sourceIsFriendly: true,
		targetID: 2,
		targetIsFriendly: false,
		ability: {
			name: 'Wildfire',
			guid: 1000861,
			type: 1,
			abilityIcon: '013000-013011.png',
		},
		stack: 2,
	},
	removebuffstack: {
		timestamp: 7410865,
		type: 'removebuffstack',
		sourceID: 7,
		sourceIsFriendly: true,
		targetID: 7,
		targetIsFriendly: true,
		ability: {
			name: 'Acceleration',
			guid: 1001238,
			type: 1,
			abilityIcon: '019000-019661.png',
		},
		stack: 1,
	},
	removedebuffstack: {
		timestamp: 431742760,
		type: 'removedebuffstack',
		sourceID: 4917,
		sourceIsFriendly: false,
		targetID: 4883,
		targetInstance: 14,
		targetIsFriendly: true,
		ability: {
			name: 'Summon Order II',
			guid: 1001964,
			type: 1,
			abilityIcon: '019000-019711.png',
		},
		stack: 1,
	},
	targetabilityupdate: {
		timestamp: 7906700,
		type: 'targetabilityupdate',
		sourceID: 20,
		sourceInstance: 7,
		sourceIsFriendly: false,
		targetID: 20,
		targetInstance: 7,
		targetIsFriendly: false,
		ability: {
			name: 'Unknown Ability',
			guid: 0,
			type: 0,
			abilityIcon: '000000-000405.png',
		},
		targetable: 1,
	},
	death: {
		timestamp: 502017,
		type: 'death',
		sourceID: 29,
		sourceInstance: 5,
		sourceIsFriendly: false,
		targetID: 8,
		targetIsFriendly: true,
		ability: {
			name: 'Unknown Ability',
			guid: 0,
			type: 0,
			abilityIcon: '000000-000405.png',
		},
	},
	limitbreakupdate: {
		...fakeBaseFields,
		bars: 3,
		timestamp: 8588751,
		type: 'limitbreakupdate',
		value: 220,
	},
	checksummismatch: {
		...fakeBaseFields,
		timestamp: 3206414,
		type: 'checksummismatch',
		data: 41298,
	},
	zonechange: {
		...fakeBaseFields,
		timestamp: 8240010,
		type: 'zonechange',
		zoneDifficulty: 0,
		zoneID: 840,
		zoneName: 'the Twinning',
	},
	unknown: {
		...fakeBaseFields,
		timestamp: 760647,
		type: 'unknown',
		ability: {
			abilityIcon: '002000-002629.png',
			guid: 139,
			name: 'Holy',
			type: 1024,
		},
	},
	dispel: {
		timestamp: 67561238,
		type: 'dispel',
		sourceID: 20,
		sourceInstance: 1,
		sourceIsFriendly: false,
		targetID: 806,
		targetIsFriendly: true,
		ability: {
			name: 'Fifth Element',
			guid: 14275,
			type: 1024,
			abilityIcon: '000000-000405.png',
		},
		extraAbility: {
			name: 'Third Eye',
			guid: 1001232,
			type: 8,
			abilityIcon: '013000-013307.png',
		},
		isBuff: false,
	},
	wipecalled: {
		timestamp: 100000000,
		type: 'wipecalled',
		sourceIsFriendly: true,
		targetIsFriendly: true,
	},
}

// #endregion

describe('Event adapter', () => {
	describe('individual events', () => {
		// Noop the reassign unknown actor step - the test data for individual events is intentionally
		// pulled directly out of real logcs, and as such will always be misaligned from the test report.
		class NoopStep extends AdapterStep {}
		beforeEach(() => {
			MockReassignUnknownActorStep.mockImplementationOnce((...args: ReassignUnknownActorStepParams) => {
				return new NoopStep(...args) as ReassignUnknownActorStep
			})
		})

		Object.keys(fakeEvents).forEach(eventType => it(`adapts ${eventType}`, () => {
			const event = fakeEvents[eventType as keyof typeof fakeEvents]
			expect(adaptEvents(report, pull, [event])).toMatchSnapshot()
		}))
	})

	it('sorts events with identical timestamps', () => {
		const result = adaptEvents(report, pull, [
			{...fakeEvents.applybuff, timestamp: 1},
			{...fakeEvents.cast, timestamp: 1},
			{...fakeEvents.begincast, timestamp: 1},
			{...fakeEvents.death, timestamp: 1},
		])

		expect(result.map(event => event.type)).toEqual([
			'actorUpdate',
			'prepare',
			'action',
			'statusApply',
		])
	})

	it('synthesizes prepull actions', () => {
		// simulating a begincast before the event that should trip the prepull, to ensure
		// prepull is added before _all_ events, not just the start of the adapted base.
		const result = adaptEvents(report, pull, [
			fakeEvents.begincast,
			fakeEvents.calculateddamage,
		])
		expect(result.map(event => event.type)).toEqual([
			'action', // prepull synth
			'prepare', // begincast
			'snapshot', // calculateddamage
			'actorUpdate',
			'actorUpdate',
		])
	})

	it('synthesizes prepull status-applying actions', () => {
		// simulating a begincast before the event that should trip the prepull, to ensure
		// prepull is added before _all_ events, not just the start of the adapted base.
		const result = adaptEvents(report, pull, [
			{...fakeEvents.begincast, timestamp: 1},
			{
				timestamp: 2,
				type: 'applybuff',
				sourceID: 11,
				sourceInstance: 2,
				sourceIsFriendly: true,
				targetID: 4,
				targetIsFriendly: true,
				ability: {
					name: 'Raging Strikes',
					guid: 1000125,
					type: 8,
					abilityIcon: '012000-012848.png',
				},
			},
		])
		expect(result.map(event => event.type)).toEqual([
			'action', // prepull synth
			'prepare', // begincast
			'statusApply', // applybuff
		])
	})

	it('synthesizes prepull status applications', () => {
		// simulating a begincast before the event that should trip the prepull, to ensure
		// prepull is added before _all_ events, not just the start of the adapted base.
		const result = adaptEvents(report, pull, [
			fakeEvents.begincast,
			fakeEvents.removebuff,
		])
		expect(result.map(event => event.type)).toEqual([
			'action', // prepull synth
			'statusApply', // prepull synth
			'prepare', // begincast
			'statusRemove', // removebuff
		])
	})

	it('merges duplicate status data', () => {
		const statusData = 10

		const result = adaptEvents(report, pull, [{
			timestamp: 100,
			type: 'applybuff',
			sourceID: 1,
			sourceIsFriendly: true,
			targetID: 2,
			targetIsFriendly: true,
			ability: fakeAbility,
		}, {
			timestamp: 100,
			type: 'applybuffstack',
			sourceID: 1,
			sourceIsFriendly: true,
			targetID: 2,
			targetIsFriendly: true,
			ability: fakeAbility,
			stack: statusData,
		}])

		expect(result).toHaveLength(1)
		expect(result[0].type).toBe('statusApply')
		expect((result[0] as Events['statusApply']).data).toBe(statusData)
	})

	it('omits duplicate actor data', () => {
		const sharedFields = {
			...fakeHitTypeFields,
			...fakeBaseFields,
			type: 'damage',
			targetID: 1,
			sourceID: 2,
			hitType: HitType.NORMAL,
			amount: 100,
			ability: fakeAbility,
		} as const

		const sharedResources = {
			hitPoints: 100,
			maxHitPoints: 1000,
			mp: 100,
			maxMP: 10000,
			tp: 0,
			maxTP: 0,
			x: 100,
			y: 100,
			facing: 0,
		}

		const result = adaptEvents(report, pull, [{
			...sharedFields,
			timestamp: 100,
			targetResources: sharedResources,
		}, {
			...sharedFields,
			timestamp: 200,
			targetResources: {
				...sharedResources,
				hitPoints: 200,
			},
		}, {
			...sharedFields,
			timestamp: 300,
			targetResources: {
				...sharedResources,
				hitPoints: 200,
				mp: 300,
				maxMP: 15000,
				x: 150,
				y: 50,
			},
		}])

		const updates = result.filter(event => true
			&& event.type === 'actorUpdate'
			&& event.actor === '1'
		)

		expect(updates).toEqual([{
			timestamp: 100,
			type: 'actorUpdate',
			actor: '1',
			hp: {current: 100, maximum: 1000},
			mp: {current: 100, maximum: 10000},
			position: {x: 100, y: 100, bearing: 0},
		}, {
			timestamp: 200,
			type: 'actorUpdate',
			actor: '1',
			hp: {current: 200},
		}, {
			timestamp: 300,
			type: 'actorUpdate',
			actor: '1',
			mp: {current: 300, maximum: 15000},
			position: {x: 150, y: 50},
		}])
	})
})
