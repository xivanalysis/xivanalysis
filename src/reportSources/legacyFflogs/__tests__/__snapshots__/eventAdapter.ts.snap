// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event adapter individual events adapts applybuff 1`] = `
Array [
  Object {
    "source": "11:2",
    "status": 1001917,
    "target": "4",
    "timestamp": 7537469,
    "type": "statusApply",
  },
]
`;

exports[`Event adapter individual events adapts applybuffstack 1`] = `
Array [
  Object {
    "action": 7518,
    "data": 2,
    "source": "7",
    "status": 1001238,
    "target": "7",
    "timestamp": -3,
    "type": "action",
  },
  Object {
    "data": 2,
    "source": "7",
    "status": 1001238,
    "target": "7",
    "timestamp": 7410286,
    "type": "statusApply",
  },
]
`;

exports[`Event adapter individual events adapts applydebuff 1`] = `
Array [
  Object {
    "action": 2258,
    "source": "6",
    "status": 1000638,
    "target": "11",
    "timestamp": -3,
    "type": "action",
  },
  Object {
    "source": "6",
    "status": 1000638,
    "target": "11",
    "timestamp": 190177,
    "type": "statusApply",
  },
]
`;

exports[`Event adapter individual events adapts applydebuffstack 1`] = `
Array [
  Object {
    "data": 2,
    "source": "8",
    "status": 1000861,
    "target": "2",
    "timestamp": 7425207,
    "type": "statusApply",
  },
]
`;

exports[`Event adapter individual events adapts begincast 1`] = `
Array [
  Object {
    "action": 16541,
    "source": "9",
    "target": "2",
    "timestamp": 7446878,
    "type": "prepare",
  },
]
`;

exports[`Event adapter individual events adapts calculateddamage 1`] = `
Array [
  Object {
    "action": 3554,
    "sequence": 16647,
    "source": "1",
    "target": "3",
    "timestamp": -1,
    "type": "action",
  },
  Object {
    "action": 3554,
    "sequence": 16647,
    "source": "1",
    "target": "3",
    "timestamp": 8011183,
    "type": "snapshot",
  },
  Object {
    "actor": "3",
    "hp": Object {
      "current": 67180802,
      "maximum": 75722720,
    },
    "mp": Object {
      "current": 10000,
      "maximum": 10000,
    },
    "position": Object {
      "bearing": -160,
      "x": 9999,
      "y": 10081,
    },
    "timestamp": 8011183,
    "type": "actorUpdate",
  },
  Object {
    "actor": "1",
    "hp": Object {
      "current": 142411,
      "maximum": 142411,
    },
    "mp": Object {
      "current": 10000,
      "maximum": 10000,
    },
    "position": Object {
      "bearing": -245,
      "x": 9260,
      "y": 10463,
    },
    "timestamp": 8011183,
    "type": "actorUpdate",
  },
]
`;

exports[`Event adapter individual events adapts calculatedheal 1`] = `
Array [
  Object {
    "action": 3571,
    "sequence": 4570,
    "source": "3",
    "target": "3",
    "timestamp": -1,
    "type": "action",
  },
  Object {
    "action": 3571,
    "sequence": 4570,
    "source": "3",
    "target": "3",
    "timestamp": 7412071,
    "type": "snapshot",
  },
  Object {
    "actor": "3",
    "hp": Object {
      "current": 122214,
      "maximum": 122214,
    },
    "mp": Object {
      "current": 9800,
      "maximum": 10000,
    },
    "position": Object {
      "bearing": -208,
      "x": 9887,
      "y": 10642,
    },
    "timestamp": 7412071,
    "type": "actorUpdate",
  },
]
`;

exports[`Event adapter individual events adapts cast 1`] = `
Array [
  Object {
    "action": 16139,
    "source": "4",
    "target": "2",
    "timestamp": 7537601,
    "type": "action",
  },
]
`;

exports[`Event adapter individual events adapts checksummismatch 1`] = `Array []`;

exports[`Event adapter individual events adapts damage 1`] = `
Array [
  Object {
    "amount": 4176,
    "cause": Object {
      "status": 1000725,
      "type": "status",
    },
    "overkill": 0,
    "sequence": undefined,
    "source": "2",
    "sourceModifier": 0,
    "target": "3",
    "targetModifier": 0,
    "timestamp": 8012166,
    "type": "damage",
  },
  Object {
    "actor": "3",
    "hp": Object {
      "current": 66978953,
      "maximum": 75722720,
    },
    "mp": Object {
      "current": 10000,
      "maximum": 10000,
    },
    "position": Object {
      "bearing": -160,
      "x": 9999,
      "y": 10081,
    },
    "timestamp": 8012166,
    "type": "actorUpdate",
  },
]
`;

exports[`Event adapter individual events adapts death 1`] = `
Array [
  Object {
    "actor": "8",
    "hp": Object {
      "current": 0,
    },
    "mp": Object {
      "current": 0,
    },
    "timestamp": 502017,
    "type": "actorUpdate",
  },
]
`;

exports[`Event adapter individual events adapts dispel 1`] = `Array []`;

exports[`Event adapter individual events adapts encounterend 1`] = `Array []`;

exports[`Event adapter individual events adapts encounterstart 1`] = `Array []`;

exports[`Event adapter individual events adapts heal 1`] = `
Array [
  Object {
    "amount": 4588,
    "cause": Object {
      "action": 500001,
      "type": "action",
    },
    "overheal": 0,
    "sequence": undefined,
    "source": "11",
    "sourceModifier": 0,
    "target": "8",
    "timestamp": 8586700,
    "type": "heal",
  },
  Object {
    "actor": "8",
    "hp": Object {
      "current": 60545,
      "maximum": 128026,
    },
    "mp": Object {
      "current": 9461,
      "maximum": 10000,
    },
    "position": Object {
      "bearing": -288,
      "x": 10601,
      "y": 9971,
    },
    "timestamp": 8586700,
    "type": "actorUpdate",
  },
]
`;

exports[`Event adapter individual events adapts limitbreakupdate 1`] = `Array []`;

exports[`Event adapter individual events adapts refreshbuff 1`] = `
Array [
  Object {
    "source": "3",
    "status": 1001873,
    "target": "3",
    "timestamp": 7416935,
    "type": "statusApply",
  },
]
`;

exports[`Event adapter individual events adapts refreshdebuff 1`] = `
Array [
  Object {
    "source": "8",
    "status": 1000861,
    "target": "2",
    "timestamp": 7423728,
    "type": "statusApply",
  },
]
`;

exports[`Event adapter individual events adapts removebuff 1`] = `
Array [
  Object {
    "action": 3539,
    "source": "1",
    "status": 1001902,
    "target": "1",
    "timestamp": -3,
    "type": "action",
  },
  Object {
    "source": "1",
    "status": 1001902,
    "target": "1",
    "timestamp": -2,
    "type": "statusApply",
  },
  Object {
    "source": "1",
    "status": 1001902,
    "target": "1",
    "timestamp": 7735462,
    "type": "statusRemove",
  },
]
`;

exports[`Event adapter individual events adapts removebuffstack 1`] = `
Array [
  Object {
    "action": 7518,
    "data": 1,
    "source": "7",
    "status": 1001238,
    "target": "7",
    "timestamp": -3,
    "type": "action",
  },
  Object {
    "data": 1,
    "source": "7",
    "status": 1001238,
    "target": "7",
    "timestamp": 7410865,
    "type": "statusApply",
  },
]
`;

exports[`Event adapter individual events adapts removedebuff 1`] = `
Array [
  Object {
    "action": 7436,
    "source": "9",
    "status": 1001221,
    "target": "2",
    "timestamp": -3,
    "type": "action",
  },
  Object {
    "source": "9",
    "status": 1001221,
    "target": "2",
    "timestamp": -2,
    "type": "statusApply",
  },
  Object {
    "source": "9",
    "status": 1001221,
    "target": "2",
    "timestamp": 7427662,
    "type": "statusRemove",
  },
]
`;

exports[`Event adapter individual events adapts removedebuffstack 1`] = `
Array [
  Object {
    "data": 1,
    "source": "4917",
    "status": 1001964,
    "target": "4883:14",
    "timestamp": 431742760,
    "type": "statusApply",
  },
]
`;

exports[`Event adapter individual events adapts targetabilityupdate 1`] = `
Array [
  Object {
    "actor": "20:7",
    "targetable": true,
    "timestamp": 7906700,
    "type": "actorUpdate",
  },
]
`;

exports[`Event adapter individual events adapts unknown 1`] = `Array []`;

exports[`Event adapter individual events adapts wipecalled 1`] = `Array []`;

exports[`Event adapter individual events adapts zonechange 1`] = `Array []`;
